File: 0024_Swap_Nodes_In_Pair/0024_Swap_Nodes_In_Pair.cc
/*
Given a linked list, swap every two adjacent nodes and return its head.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // different way with python method, use every two nodes, only change value
    ListNode* swapPairs(ListNode* head) {
        if(head==NULL||head->next==NULL)
            return head;
        ListNode *first=head, *second = head->next,*temp = head->next->next;
        second->next = first;
        first->next = swapPairs(temp);
        return second;
    }
    /*ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        ListNode dummy  = ListNode(0, head);
        ListNode* prev = &dummy;
        ListNode* cur = head;
        while (cur && cur->next) {
            ListNode* next = cur->next;
            ListNode* nextPair = next->next;
            next->next = cur;
            cur->next = nextPair;
            
            prev->next = next;
            prev = cur;
            cur = nextPair;
        }
        return dummy.next;
    }*/
};

/*
Example 1:
Input: head = [1,2,3,4]
Output: [2,1,4,3]

Example 2:
Input: head = []
Output: []

Example 3:
Input: head = [1]
Output: [1]

Time: O(N)
*/

File: 0024_Swap_Nodes_In_Pair/0024_Swap_Nodes_In_Pair.py
"""

"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None:
            return None
        if head.next == None:
            return head
        
        dummy = ListNode(0, head)
        prev_node = dummy
        cur_node = dummy.next
        next_node = cur_node.next
        while cur_node and next_node:
            # swap the pair
            next_pair = next_node.next
            cur_node.next = next_pair
            next_node.next = cur_node
            
            # change what prev points to assure the order,
            # since next_node is ahead of cur_node
            # prev next should be next_node istead of cur_node
            prev_node.next = next_node
            
            # update nodes
            prev_node = cur_node
            cur_node = next_pair
            next_node = cur_node.next if cur_node else None
            
        return dummy.next

"""
Example 1:
Input: head = [1,2,3,4]
Output: [2,1,4,3]

Example 2:
Input: head = []
Output: []

Example 3:
Input: head = [1]
Output: [1]

Time: O(N)
"""
